packages/pay/src/request.ts

```diff
@@ -0,0 +1,396 @@
+
+export interface Plugin {
+  sendData(data: Buffer): Promise<Buffer>
+}
+
+/** Generate keys and function to send ILP Prepares over STREAM */
+export const generateKeys = async (
+  plugin: Plugin,
+  sharedSecret: Buffer
+): Promise<(request: StreamRequest) => Promise<StreamReply>> => {
+  const encryptionKey = await generatePskEncryptionKey(sharedSecret)
+  const fulfillmentKey = await generateFulfillmentKey(sharedSecret)
+
+  return async (request: StreamRequest): Promise<StreamReply> => {
+    // Create the STREAM request packet
+    const {
+      sequence,
+      sourceAmount,
+      destinationAddress,
+      minDestinationAmount,
+      frames,
+      isFulfillable,
+      expiresAt,
+      log,
+    } = request
+
+    const streamRequest = new Packet(
+      sequence,
+      IlpPacketType.Prepare.valueOf(),
+      minDestinationAmount.toLong(),
+      frames
+    )
+
+    const data = await streamRequest.serializeAndEncrypt(encryptionKey)
+
+    let executionCondition: Buffer
+    let fulfillment: Buffer | undefined
+
+    if (isFulfillable) {
+      fulfillment = await generateFulfillment(fulfillmentKey, data)
+      executionCondition = await hash(fulfillment)
+      log.debug(
+        'sending Prepare. amount=%s minDestinationAmount=%s frames=[%s]',
+        sourceAmount,
+        minDestinationAmount,
+        frames.map((f) => FrameType[f.type]).join()
+      )
+    } else {
+      executionCondition = generateRandomCondition()
+      log.debug(
+        'sending unfulfillable Prepare. amount=%s frames=[%s]',
+        sourceAmount,
+        frames.map((f) => FrameType[f.type]).join()
+      )
+    }
+
+    log.trace('loading Prepare with frames: %o', frames)
+
+    // Create and serialize the ILP Prepare
+    const preparePacket = serializeIlpPrepare({
+      destination: destinationAddress,
+      amount: sourceAmount.toString(), // Max packet amount controller always limits this to U64
+      executionCondition,
+      expiresAt,
+      data,
+    })
+
+    // Send the packet!
+    const pendingReply = plugin
+      .sendData(preparePacket)
+      .then((data) => {
+        try {
+          return deserializeIlpReply(data)
+        } catch (_) {
+          return createReject(IlpError.F01_INVALID_PACKET)
+        }
+      })
+      .catch((err) => {
+        log.error('failed to send Prepare:', err)
+        return createReject(IlpError.T00_INTERNAL_ERROR)
+      })
+      .then((ilpReply) => {
+        if (!isFulfill(ilpReply) || !fulfillment || ilpReply.fulfillment.equals(fulfillment)) {
+          return ilpReply
+        }
+
+        log.error(
+          'got invalid fulfillment: %h. expected: %h, condition: %h',
+          ilpReply.fulfillment,
+          fulfillment,
+          executionCondition
+        )
+        return createReject(IlpError.F05_WRONG_CONDITION)
+      })
+
+    // Await reply and timeout if the packet expires
+    const timeoutDuration = expiresAt.getTime() - Date.now()
```

Comment:
I removed it because I think this only matters for nodes handling _incoming_ ILP Prepares.

With a timeout longer than the `expiresAt`, the behavior seems the same as if it just set a longer `expiresAt`, so deducting a duration didn't seem to be doing anything.

By contrast, if a node is either fulfilling packets or returning a Fulfill, then it cares that there's enough time from when it gets the reply, that it can still deliver the reply back to the previous node. It's a little arbitrary but I think the onus is the next hop node to get the Fulfill back in time, not the previous hop.